easybuild/easyblocks/generic/cargo.py

```diff
@@ -392,44 +388,72 @@ def _setup_offline_config(self, git_sources):
+                        append=True
+                    )
                 else:
-                    self.log.debug("Preparing config.toml entry for git repo: %s", git_repo)
-                    # Fallback for (older) easyconfigs that don't specify the branch:
-                    # Patch the whole repo, which is only possible if there is a single revision used.
-                    config_toml_git_patches[git_repo] += CONFIG_TOML_PATCH_GIT_CRATES.format(
-                        crate=src['crate_name'], path=src_dir)
+                    self.log.debug(f"Writing config.toml entry for git repo: {git_repo} , rev {rev}")
+                    write_file(
+                        config_toml,
+                        CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev),
+                        append=True
+                    )
             else:
                 self.log.debug("Writing config.toml entry for git repo: %s rev %s", git_repo, rev)
                 # Workspace sources stay in their own separate folder.
                 # We cannot have a `directory = "<dir>"` entry where a folder containing a workspace is inside
                 write_file(config_toml,
                            CONFIG_TOML_SOURCE_GIT_WORKSPACE.format(url=git_repo, rev=rev, workspace_dir=src_dir),
                            append=True)
-        for git_repo, config_crates in config_toml_git_patches.items():
-            self.log.debug("Writting config.toml patch entry for git repo: %s", git_repo)
-            write_file(config_toml, CONFIG_TOML_PATCH_GIT.format(repo=git_repo, crates=config_crates), append=True)
 
         # Use environment variable since it would also be passed along to builds triggered via python packages
         env.setvar('CARGO_NET_OFFLINE', 'true')
 
+    def _get_crate_git_repo_branch(self, crate_name):
+        """
+        Seek the dependency definition for given crate in all Cargo.toml files of sources
+        Return branch target for given crate_name if any
+        """
+        cargo_toml_glob = os.path.join(self.builddir, f'{self.name}-{self.version}', '**', 'Cargo.toml')
+        cargo_toml_files = glob.glob(cargo_toml_glob, recursive=True)
+        self.log.debug(
+            f"Searching definition of crate '{crate_name}' in the following files: {', '.join(cargo_toml_files)}"
+        )
+
+        if not cargo_toml_files:
+            raise EasyBuildError("Cargo.toml file not found in sources")
+
+        git_repo_spec_pattern = f"{crate_name} = ({{.*}})"
+        git_repo_spec = re.compile(git_repo_spec_pattern, re.M)
+        git_branch_spec_pattern = r'branch\s*=\s*"(.*)"'
+        git_branch_spec = re.compile(git_branch_spec_pattern, re.M)
+
+        for cargo_toml in cargo_toml_files:
+            git_repo_crate = git_repo_spec.search(read_file(cargo_toml))
+            if git_repo_crate:
+                self.log.debug(f"Found specification in Cargo.toml for crate '{crate_name}': {git_repo_crate.group()}")
```

Comment:
Shall we include the (relative) path to the Cargo.toml file in which the match was found? Also some consideration of f-string

Reply:
If you include path to `Cargo.toml`, make it full path